ydb/library/yql/minikql/comp_nodes/mkql_grace_join.cpp

```diff
@@ -566,6 +565,383 @@ TGraceJoinPacker::TGraceJoinPacker(const std::vector<TType *> & columnTypes, con
+                *PartialJoinCompleted = true;
+                LeftPacker->StartTime = std::chrono::system_clock::now();
+                RightPacker->StartTime = std::chrono::system_clock::now();
+                if ( SelfJoinSameKeys_ ) {
+                    JoinedTablePtr->Join(*LeftPacker->TablePtr, *LeftPacker->TablePtr, JoinKind, *HaveMoreLeftRows, *HaveMoreRightRows);
+                } else {
+                    JoinedTablePtr->Join(*LeftPacker->TablePtr, *RightPacker->TablePtr, JoinKind, *HaveMoreLeftRows, *HaveMoreRightRows);
+                }
+                JoinedTablePtr->ResetIterator();
+                LeftPacker->EndTime = std::chrono::system_clock::now();
+                RightPacker->EndTime = std::chrono::system_clock::now();
+            }
+
+        }
+
+        return EFetchResult::Finish;
+    }
+
+    bool TryToReduceMemory() {
+        bool isWaitingLeftForReduce = LeftPacker->TablePtr->TryToReduceMemory();
+        bool isWaitingRightForReduce = RightPacker->TablePtr->TryToReduceMemory();
+
+        return isWaitingLeftForReduce || isWaitingRightForReduce;
+    }
+
+    void UpdateSpilling() {
+        LeftPacker->TablePtr->UpdateSpilling();
+        RightPacker->TablePtr->UpdateSpilling();
+    }
+
+    bool HasRunningAsyncOperation() const {
+        return LeftPacker->TablePtr->HasRunningAsyncIoOperation() || RightPacker->TablePtr->HasRunningAsyncIoOperation();
+    }
+
+void DoCalculateWithSpilling(TComputationContext& ctx) {
+    UpdateSpilling();
+
+    if (!HasMemoryForProcessing()) {
+        bool isWaitingForReduce = TryToReduceMemory();
+        if (isWaitingForReduce) return;
+    }
+
+    while (*HaveMoreLeftRows || *HaveMoreRightRows) {
+        bool isYield = FetchAndPackData(ctx);
+        if (isYield) return;
+    }
+
+    if (!*HaveMoreLeftRows && !*HaveMoreRightRows) {
+        UpdateSpilling();
+        if (HasRunningAsyncOperation()) return;
+        if (!IsSpillingFinalized) {
+            LeftPacker->TablePtr->FinalizeSpilling();
+            RightPacker->TablePtr->FinalizeSpilling();
+            IsSpillingFinalized = true;
+
+            if (HasRunningAsyncOperation()) return;
+        }
+        SwitchMode(EOperatingMode::ProcessSpilled, ctx);
+        return;
+    }
+}
+
+EFetchResult ProcessSpilledData(TComputationContext&, NUdf::TUnboxedValue*const* output) {
+    while (NextBucketToJoin != GraceJoin::NumberOfBuckets) {
+        UpdateSpilling();
+
+        if (HasRunningAsyncOperation()) return EFetchResult::Yield;
+
+        if (!LeftPacker->TablePtr->IsBucketInMemory(NextBucketToJoin)) {
+            LeftPacker->TablePtr->StartLoadingBucket(NextBucketToJoin);
+        }
+
+        if (!RightPacker->TablePtr->IsBucketInMemory(NextBucketToJoin)) {
+            RightPacker->TablePtr->StartLoadingBucket(NextBucketToJoin);
+        } 
+
+        if (LeftPacker->TablePtr->IsBucketInMemory(NextBucketToJoin) && RightPacker->TablePtr->IsBucketInMemory(NextBucketToJoin)) {
```

Comment:
I think this condition should always be true, because before that we already checked that they are not false.

Reply:
Previous check starts loading buckets if they are not in memory. 

On this line buckets may be in memory (if they were emoty and loaded immediately) or not in memory (in this case we will yield)